cw_decred/lib/wallet.dart

```diff
@@ -0,0 +1,604 @@
+
+    final syncStatusCode = decoded["syncstatuscode"] ?? 0;
+    final syncStatusStr = decoded["syncstatus"] ?? "";
+    final targetHeight = decoded["targetheight"] ?? 1;
+    final numPeers = decoded["numpeers"] ?? 0;
+    // final cFiltersHeight = decoded["cfiltersheight"] ?? 0;
+    final headersHeight = decoded["headersheight"] ?? 0;
+    final rescanHeight = decoded["rescanheight"] ?? 0;
+
+    if (numPeers == 0) {
+      syncStatus = NotConnectedSyncStatus();
+      return false;
+    }
+
+    // Sync codes:
+    // NotStarted = 0
+    // FetchingCFilters = 1
+    // FetchingHeaders = 2
+    // DiscoveringAddrs = 3
+    // Rescanning = 4
+    // Complete = 5
+
+    if (syncStatusCode > 4) {
+      syncStatus = SyncedSyncStatus();
+      // Initiate a receive address in case we lose peers later.
+      if (walletAddresses.currentAddr == '') {
+        walletAddresses.address;
+      }
+      return true;
+    }
+
+    if (syncStatusCode == 0) {
+      syncStatus = ConnectedSyncStatus();
+      return false;
+    }
+
+    if (syncStatusCode == 1) {
+      syncStatus = SyncingSyncStatus(targetHeight, 0.0);
+      return false;
+    }
+
+    if (syncStatusCode == 2) {
+      final headersProg = headersHeight / targetHeight;
+      // Only allow headers progress to go up half way.
+      syncStatus =
+          SyncingSyncStatus(targetHeight - headersHeight, headersProg / 2);
+      return false;
+    }
+
+    // TODO: This step takes a while so should really get more info to the UI
+    // that we are discovering addresses.
+    if (syncStatusCode == 3) {
+      // Hover at half.
+      syncStatus = SyncingSyncStatus(0, .5);
+      return false;
+    }
+
+    if (syncStatusCode == 4) {
+      // Start at 75%.
+      final rescanProg = rescanHeight / targetHeight / 4;
+      syncStatus =
+          SyncingSyncStatus(targetHeight - rescanHeight, .75 + rescanProg);
+      return false;
+    }
+    return false;
+  }
+
+  @action
+  @override
+  Future<void> connectToNode({required Node node}) async {
+    if (connecting) {
+      throw "decred already connecting";
+    }
+    connecting = true;
+    String addr = "";
+    if (node.uri.host != "") {
+      addr = node.uri.host;
+      if (node.uri.port != "") {
+        addr += ":" + node.uri.port.toString();
+      }
+    }
+    if (addr != persistantPeer) {
+      if (syncTimer != null) {
+        syncTimer!.cancel();
+        syncTimer = null;
+      }
+      persistantPeer = addr;
+      libdcrwallet.closeWallet(walletInfo.name);
+      final network = walletInfo.derivationPath ==
+                  DecredWalletService.seedRestorePathTestnet ||
+              walletInfo.derivationPath ==
+                  DecredWalletService.pubkeyRestorePathTestnet
+          ? "testnet"
+          : "mainnet";
+      libdcrwallet.loadWalletSync({
+        "name": walletInfo.name,
+        "dataDir": walletInfo.dirPath,
+        "network": network,
+      });
+    }
```

Comment:
The wallet api only allows starting sync, I don't think we can stop it. And we also don't have a command to set peers with the wallet atm. So, need to stop the wallet to stop sync and then start sync again. This could be better but may require changes in dcrwallet. I'll look again to make sure.